Epic 5886 : ODBC for hosted clients

labkey-bpatel opened the pull request:
#### Rationale
Part of offering ODBC connectivity to our hosted customers is to give end-user access to the configured certificate.

#### Related Pull Requests
https://github.com/LabKey/connectors/pull/34

#### Changes
- Add External Tool View Service and Provider infrastructure
- Decouple APIKey view from SecurityController.ApiKeyAction to APIKeyProvider.
- Refactor ApiKeyAction to iterate through External Tool Providers.
- Decouple user navigation piece ('Done' button) from the individual view rendering.

labkey-bpatel committed:
Add External Tool View Service and Provider infrastructure.
- Decouple APIKey view from SecurityController.ApiKeyAction to APIKeyProvider.
- Refactor ApiKeyAction to iterate through External Tool Providers.
- Decouple user navigation piece ('Done' button) from the individual view rendering.
- Add method to check for ReadPermission.

labkey-bpatel committed:
Add header to APIKey view. Optimize imports.

labkey-bpatel committed:
Add line breaks.

labkey-bpatel committed:
Merge remote-tracking branch 'origin/develop' into fb_odbcSelfSignedCertificate

labkey-jeckels commented on core/src/org/labkey/core/security/SecurityController.java:
```diff
@@ -1928,17 +1930,29 @@ public void setHideCaption(String hideCaption)
+                }
+            }
+
+            //TODO: this is no longer shown, should this message be part of the view or silently just not display because user does not have the access?
```
Let's replace with a "Nothing enabled" page if somebody gets here but no providers are active for them

labkey-bpatel replied:
Done!

labkey-jeckels reviewed (approved):
Let me know if you want further review, but otherwise one simple comment and don't forget the TODOs about JavaDocs

Commit:
Merge in changes from develop

labkey-bpatel committed:
Code Review changes:
- Add javadocs.
- Add a "nothing enabled" page, if there are not views from the providers.

labkey-jeckels commented on core/src/org/labkey/core/security/SecurityController.java:
```diff
@@ -1939,9 +1939,10 @@ public ModelAndView getView(ReturnUrlForm form, BindException errors)
-            //TODO: this is no longer shown, should this message be part of the view or silently just not display because user does not have the access?
-//            if (!PopupUserView.allowApiKeyPage(getUser()))
-//                throw new UnauthorizedException("API keys are not configured on this site. Contact a site administrator.");
+            if (view.isEmpty())
```
We should rename this action - it's no longer just for ApiKeys

labkey-bpatel replied:
Yes, good point! Action now renamed.

labkey-jeckels commented on core/src/org/labkey/core/security/nothingEnabled.jsp:
```diff
@@ -0,0 +1,20 @@
+%>
+<%@ page extends="org.labkey.api.jsp.JspBase" %>
+<%@ taglib prefix="labkey" uri="http://www.labkey.org/taglib" %>
+Nothing enabled
```
Sorry, I didn't literally mean to just show "nothing enabled". Let's rephrase to be a bit more user-friendly. How about:

This server does not have any configurations enabled that show settings here. An administrator can configure tool access via the Admin Console.

labkey-bpatel replied:
Haha yes, sorry i meant to ask you if we needed anything more user-friendly. Thanks for the text.

Commit:
Revert accidental commit

labkey-jeckels commented on api/src/org/labkey/api/view/PopupUserView.java:
```diff
@@ -84,4 +85,9 @@ public static boolean allowApiKeyPage(User user)
+
+    public static boolean allowExternalToolSettingsPage(User user)
+    {
+        return getContextContainer().hasPermission(user, ReadPermission.class);
```
Is this quite right? Seems like the current user will always have read permission on the current container if the page is being rendered, and if ODBC is off and API keys aren't enabled, they'll get the page that tells them nothing is available.

Seems like we should check to see if any views would be shown to the user, and if not, omit the menu item.

labkey-bpatel replied:
Since all the readers get odbc view at minimum, not too clear what the check should be to omit the menu item altogether, (and if odbc is disabled then its just not configured and so will get nothingEnabled page). So not clear on what the criteria is to for a user not to get external tool provider views.

labkey-jeckels replied:
Not all deployments will have the Connectors module, and API keys aren't enabled by default.

There's no harm in showing them a blank page, of course, but it's a useless bit of menu clutter in those cases.

labkey-bpatel replied:
Done, let me know what you think!

labkey-jeckels reviewed (changes requested):
Another set of suggestions, hopefully the last on this side of the PR

Commit:
Merge remote-tracking branch 'origin/fb_odbcSelfSignedCertificate' into fb_odbcSelfSignedCertificate

labkey-bpatel committed:
Code Review changes:
- Rename APIKeyAction to ExternalToolsViewAction, and related refactor.
- Add more user-friendly text for nothingEnabled page.

labkey-bpatel committed:
Code Review changes:
- Remove check for allowExternalToolSettingsPage() - not effective.
- Other minor updates.

labkey-bpatel committed:
Code Review change - Omit External Tool Setting menu item if views are not available to the user.

labkey-jeckels commented on core/src/org/labkey/core/security/SecurityController.java:
```diff
@@ -269,12 +271,17 @@ public ActionURL getFolderAccessURL(Container container)
 
         @Override
-        public ActionURL getApiKeyURL(@NotNull URLHelper returnURL)
+        public ActionURL getExternalToolsViewURL(User user, @NotNull URLHelper returnURL)
```
Let's annotate this as @Nullable, and consider this set!

labkey-jeckels reviewed (approved):
One more tiny tweak, but otherwise ready to merge so I'm approving the PR

labkey-bpatel committed:
Annotate with @Nullable

labkey-bpatel committed:
Merge remote-tracking branch 'origin/develop' into fb_odbcSelfSignedCertificate

labkey-bpatel committed:
Use a more reliable approach to identifying External Tools Setting views.

labkey-bpatel committed:
Merge remote-tracking branch 'origin/develop' into fb_odbcSelfSignedCertificate
